actors/builtin/power/power_test.go

```diff
@@ -72,6 +72,243 @@ func TestConstruction(t *testing.T) {
+		require.EqualValues(t, p1, evt.CallbackPayload)
+		require.EqualValues(t, miner, evt.MinerAddr)
+
+		found, err = events.Get(1, &evt)
+		require.NoError(t, err)
+		require.True(t, found)
+		require.EqualValues(t, p2, evt.CallbackPayload)
+		require.EqualValues(t, miner, evt.MinerAddr)
+
+		// enroll another event with a different miner for a different epoch
+		e2 := abi.ChainEpoch(2)
+		p3 := []byte("test")
+		miner2 := tutil.NewIDAddr(t, 501)
+		ac.enrollCronEvent(rt, miner2, e2, p3)
+		events = ac.getEnrolledCronTicks(rt, e2)
+		found, err = events.Get(0, &evt)
+		require.NoError(t, err)
+		require.True(t, found)
+		require.EqualValues(t, p3, evt.CallbackPayload)
+		require.EqualValues(t, miner2, evt.MinerAddr)
+	})
+
+	t.Run("fails if epoch is negative", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			ac.enrollCronEvent(rt, miner, abi.ChainEpoch(-1), []byte("payload"))
+		})
+	})
+}
+
+func TestOnConsensusFault(t *testing.T) {
+	miner := tutil.NewIDAddr(t, 101)
+	owner := tutil.NewIDAddr(t, 102)
+	smallPowerUnit := big.NewInt(1_000_000)
+	powerUnit := power.ConsensusMinerMinPower
+	zeroPledge := abi.NewTokenAmount(0)
+
+	t.Run("qaPower was below threshold before fault", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		ac.createMinerBasic(rt, owner, owner, miner)
+		ac.updateClaimedPower(rt, miner, smallPowerUnit, smallPowerUnit)
+
+		st := getState(rt)
+		require.True(t, st.TotalRawBytePower.IsZero())
+		require.True(t, st.TotalQualityAdjPower.IsZero())
+		require.EqualValues(t, 0, st.MinerAboveMinPowerCount)
+
+		ac.onConsensusFault(rt, miner, &zeroPledge)
+
+		st = getState(rt)
+		require.True(t, st.TotalRawBytePower.IsZero())
+		require.True(t, st.TotalQualityAdjPower.IsZero())
+		require.EqualValues(t, 0, st.MinerAboveMinPowerCount)
+	})
+
+	t.Run("qaPower was above threshold before fault", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		ac.createMinerBasic(rt, owner, owner, miner)
+		ac.updateClaimedPower(rt, miner, powerUnit, powerUnit)
+
+		st := getState(rt)
+		require.EqualValues(t, powerUnit, st.TotalRawBytePower)
+		require.EqualValues(t, powerUnit, st.TotalQualityAdjPower)
+		require.EqualValues(t, 1, st.MinerAboveMinPowerCount)
+
+		ac.onConsensusFault(rt, miner, &zeroPledge)
+
+		st = getState(rt)
+		require.True(t, st.TotalRawBytePower.IsZero())
+		require.True(t, st.TotalQualityAdjPower.IsZero())
+		require.EqualValues(t, 0, st.MinerAboveMinPowerCount)
+	})
+
+	t.Run("successfully reduces pledged amount", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		ac.createMinerBasic(rt, owner, owner, miner)
+		ac.updateClaimedPower(rt, miner, powerUnit, powerUnit)
+
+		delta := abi.NewTokenAmount(100)
+		ac.updatePledgeTotal(rt, miner, delta)
+
+		slash := abi.NewTokenAmount(50)
+		ac.onConsensusFault(rt, miner, &slash)
+
+		st := getState(rt)
+		require.EqualValues(t, big.Sub(delta, slash), st.TotalPledgeCollateral)
+	})
```

Comment:
I suggest folding this into the tests above.

Reply:
This is done.